# PC1 - Grupo 2 (Python)

#### Integrantes
- GARCIA RODRIGUEZ, EMILIO ALONSO
- PADILLA AQUISE, ALESSANDRO PIERO
- RIEGA NUÑEZ, GABRIEL ANTONIO FERMIN
- SALAMANCA FERNANDEZ, LUCAS PABLO
- SILVA ANDUJAR, NICOLAS

## 1. Data analysis

In [56]:
import numpy as np
import pandas as pd

In [57]:
df = pd.read_csv("wage2015_subsample_inference.csv")

1.Import the data set. Make sure the column names are imported as intended.d variable

In [58]:
df.columns

Index(['rownames', 'wage', 'lwage', 'sex', 'shs', 'hsg', 'scl', 'clg', 'ad',
       'mw', 'so', 'we', 'ne', 'exp1', 'exp2', 'exp3', 'exp4', 'occ', 'occ2',
       'ind', 'ind2'],
      dtype='object')

In [59]:
df = df.set_index('rownames')

2.Are there missing values? Display the number of missings (if any) of each variable in the data set.

In [60]:
missing_values = df.isnull().sum()

print(missing_values)
print("There are no missing values")

wage     0
lwage    0
sex      0
shs      0
hsg      0
scl      0
clg      0
ad       0
mw       0
so       0
we       0
ne       0
exp1     0
exp2     0
exp3     0
exp4     0
occ      0
occ2     0
ind      0
ind2     0
dtype: int64
There are no missing values


3.Report descriptive statistics of the variables (mean, standard deviation, percentiles, etc.). Interpret your results.

In [61]:
df.describe()


wage        lwage          sex          shs          hsg  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean     23.410410     2.970787     0.444466     0.023301     0.243883   
std      21.003016     0.570385     0.496955     0.150872     0.429465   
min       3.021978     1.105912     0.000000     0.000000     0.000000   
25%      13.461538     2.599837     0.000000     0.000000     0.000000   
50%      19.230769     2.956512     0.000000     0.000000     0.000000   
75%      27.777778     3.324236     1.000000     0.000000     0.000000   
max     528.845673     6.270697     1.000000     1.000000     1.000000   

               scl          clg           ad           mw           so  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean      0.278058     0.317670     0.137087     0.259612     0.296505   
std       0.448086     0.465616     0.343973     0.438464     0.456761   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     1.000000     0.000000     1.000000     1.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

                we           ne         exp1         exp2         exp3  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean      0.216117     0.227767    13.760583     3.018925     8.235867   
std       0.411635     0.419432    10.609465     4.000904    14.488962   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     5.000000     0.250000     0.125000   
50%       0.000000     0.000000    10.000000     1.000000     1.000000   
75%       0.000000     0.000000    21.000000     4.410000     9.261000   
max       1.000000     1.000000    47.000000    22.090000   103.823000   

              exp4            occ         occ2            ind         ind2  
count  5150.000000    5150.000000  5150.000000    5150.000000  5150.000000  
mean     25.118038    5310.737476    11.670874    6629.154951    13.316893  
std      53.530225   11874.356080     6.966684    5333.443992     5.701019  
min       0.000000      10.000000     1.000000     370.000000     2.000000  
25%       0.062500    1740.000000     5.000000    4880.000000     9.000000  
50%       1.000000    4040.000000    13.000000    7370.000000    14.000000  
75%      19.448100    5610.000000    17.000000    8190.000000    18.000000  
max     487.968100  100000.000000    22.000000  100000.000000    22.000000

In [62]:
df.median()

wage       19.230769
lwage       2.956512
sex         0.000000
shs         0.000000
hsg         0.000000
scl         0.000000
clg         0.000000
ad          0.000000
mw          0.000000
so          0.000000
we          0.000000
ne          0.000000
exp1       10.000000
exp2        1.000000
exp3        1.000000
exp4        1.000000
occ      4040.000000
occ2       13.000000
ind      7370.000000
ind2       14.000000
dtype: float64

In [63]:
df.mode()

wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we   ne  \
0  19.230769  2.956512  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

   exp1  exp2   exp3    exp4     occ  occ2    ind  ind2  
0   5.0  0.25  0.125  0.0625  4700.0    17  770.0    18

#### wage:
The average wage in the sample is 23.41, with a standard deviation of 21.03, reflecting significant variability in earnings. Wages range from a minimum of 3.02 to a maximum of 528.85, indicating a broad spectrum of income levels. The bottom 25th percentile earn less than 13.46, while the median wage is 19.23, and the top 25% earn more than 27.78. The fact that the median is 19.23 suggests the presence of a number of extremely high wage values, which likely skews the distribution.

#### lwage:
The logarithm of wage has a mean of 2.97 and a standard deviation of 0.57, suggesting that the distribution of wages in its logarithmic scale is more symmetric than in its original form. The logarithmic wage values range from a minimum of 1.11 to a maximum of 6.27, with the top 25% of the sample having a log wage greater than 3.32. Also, it has a median of 2.96, that suggest a lower presence of extremely high or low wages in the sample. 

#### sex:
Although sex is a dummy variable, we can see that 44% of the sample are females, while the mode is "male." The standard deviation of 0.5 suggests an almost balanced distribution between males and females.

#### shs:
The "some high school" variable shows that only 2.3% of the sample have some incomplete high school education, which represents a small fraction of the total population. 

#### hsg:
About 24% of the sample has graduated from high school, according to the mean of this variable. The standard deviation of 0.43 reflects moderate diversity in the education levels within the sample, with a significant portion not having attained a college degree. 

#### scl:
Approximately 28% of the respondents have attended some college without necessarily graduating, as indicated by the mean. This suggests that a considerable segment of the population has received some tertiary education, although they may not have completed their studies.

#### clg:
The data shows that 32% of the sample holds a college degree, as reflected in the mean of this variable. This implies that nearly one-third of the sample has completed a college education, which is significant for evaluating human capital.

#### ad:
About 14% of the individuals in the sample have an advanced degree (master's or doctorate), according to the mean of this variable. This relatively low percentage highlights that advanced education is less common compared to basic college education.

#### mw:
The Midwest region represents 26% of the sample, as indicated by the mean of this variable. This suggests that this region has significant representation within the dataset.

#### so:
The South is the most represented region, with 30% of the sample residing in this area, as shown by the mean of this variable. The higher presence of the South may reflect certain geographical trends in the sample.

#### we:
The West region accounts for 22% of the sample, according to the mean. This proportion suggests a relatively equitable distribution among the country's regions, although with less representation compared to the South.

#### ne:
The Northeast region represents 23% of the sample, as indicated by the mean of this variable. This proportion, similar to the other regions, shows a relatively balanced geographical distribution in the dataset.

#### exp1:
The average work experience is 13.76 years, with a standard deviation of 10.61, indicating high variability in the number of years worked among the individuals in the sample. The median is 10 years, meaning that half of the sample has 10 years or less of work experience, while the maximum is 47 years. Also, the mode of work experience is 5.

#### exp2:
The first polynomial of work experience has a mean of 3.02 years, with a standard deviation of 4.00. This suggests that, for certain specific aspects of work experience, most of the sample has relatively low experience.

#### exp3:
The second polynomial of work experience, exp3, has a mean of 8.24 years and a standard deviation of 14.49, indicating a wide variability.

#### exp4:
The third polynomial of work experience, exp4, shows a significantly higher mean of 25.12 years, with a standard deviation of 53.53, again reflecting a high variability.

4.How many women with a college graduate degree (clg) or above have a wage corresponding to the 25% richest of the sample? Report the dataframe corresponding to this conditions and the result.



In [64]:
percentile_75_wage = df['wage'].quantile(0.75)
rich_ed_women = df[(df['sex'] == 1) & 
                          ((df['clg'] == 1) | (df['ad'] == 1)) & 
                          (df['wage'] >= percentile_75_wage)]

print(rich_ed_women)

                wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we  \
rownames                                                                      
19         28.846154  3.361977  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
191        42.307692  3.744969  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   
232        41.208791  3.718652  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
319       100.000000  4.605170  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   
563        33.653846  3.516127  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
...              ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
32251      48.076923  3.872802  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   
32314      30.145530  3.406037  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
32419      29.914530  3.398344  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
32567      48.076923  3.872802  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
32596      45.546559  3.818735  1.0  0.0  0.0  0.0  

In [65]:
num_rich_ed_women = rich_ed_women.shape[0]
print(f"Number of women with a college graduate degree or above in the top 25% richest: {num_rich_ed_women}")

Number of women with a college graduate degree or above in the top 25% richest: 419


5.How many men with a high school graduate degree (hsg) or below have a wage corresponding to the 25% richest of the sample? Report the dataframe corresponding to this conditions and the result.

In [66]:
rich_noned_men =df[(df["sex"]== 0) &
                   ((df["hsg"]==1) | (df["shs"]==1))&
                   (df["wage"]>=percentile_75_wage)]
print(rich_noned_men)

               wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we  \
rownames                                                                     
113       27.884615  3.328075  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
276       28.846154  3.361977  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
467       28.846154  3.361977  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
858       28.846154  3.361977  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
876       29.714286  3.391628  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
...             ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
32191     33.653846  3.516127  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32254     40.865385  3.710283  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32281     31.250000  3.442019  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32321     33.653846  3.516127  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32631     32.967033  3.495508  0.0  0.0  1.0  0.0  0.0  0.0  0.0

In [67]:
num_rich_noned_men = rich_noned_men.shape[0]
print(f"Number of men with a high school graduate degree or below in the top 25% richest: {num_rich_noned_men}")

Number of men with a high school graduate degree or below in the top 25% richest: 118


6.Create two dataframes. One containing only the log(wage) and the other containig every variable of the data set but the wage related variable

In [68]:
#Dataframe only with lwage
df_logwage = df[['lwage']]
#Dataframe with rest of the variables but the wage related variable
df_nowage =df.drop(columns=['wage', 'lwage'])

## 2. Data wrangling

7.Make an array for our y 
 variable, which will be the logarithm of wage (lwage column)

In [69]:
log_w = np.array(df_logwage['lwage'])
log_w = log_w.reshape(-1, 1)

In [70]:
print(log_w)
log_w.shape

[[2.26336438]
 [3.87280229]
 [2.40312632]
 ...
 [3.64965874]
 [3.49550806]
 [2.85115104]]


(5150, 1)

8.Make three new arrays for our predictors:

8.1. The basic model will have the columns sex hsg scl clg ad so we ne exp1 occ2 ind2. Make sure to convert occ2 and ind2 to dummies and to drop the first dummy value to prevent multicolinearity. This means:

In [71]:
df_with_dummies = pd.get_dummies(df, columns=['occ2', 'ind2'], drop_first=True)
print(df_with_dummies.columns)

Index(['wage', 'lwage', 'sex', 'shs', 'hsg', 'scl', 'clg', 'ad', 'mw', 'so',
       'we', 'ne', 'exp1', 'exp2', 'exp3', 'exp4', 'occ', 'ind', 'occ2_2',
       'occ2_3', 'occ2_4', 'occ2_5', 'occ2_6', 'occ2_7', 'occ2_8', 'occ2_9',
       'occ2_10', 'occ2_11', 'occ2_12', 'occ2_13', 'occ2_14', 'occ2_15',
       'occ2_16', 'occ2_17', 'occ2_18', 'occ2_19', 'occ2_20', 'occ2_21',
       'occ2_22', 'ind2_3', 'ind2_4', 'ind2_5', 'ind2_6', 'ind2_7', 'ind2_8',
       'ind2_9', 'ind2_10', 'ind2_11', 'ind2_12', 'ind2_13', 'ind2_14',
       'ind2_15', 'ind2_16', 'ind2_17', 'ind2_18', 'ind2_19', 'ind2_20',
       'ind2_21', 'ind2_22'],
      dtype='object')


In [72]:
basic_model_vars = ['sex', 'hsg', 'scl', 'clg', 'ad', 'so', 'we', 'ne', 'exp1']
basic_model_vars += [col for col in df_with_dummies.columns if col.startswith('occ2_') or col.startswith('ind2_')]
#Create the basic model array
basic_model_array = df_with_dummies[basic_model_vars].values


basic_model_array.shape

(5150, 50)

8.2. The flexible model will have the same columns, and will also include polynomials for experience (exp2 exp3 exp4), as well as the interactions between all experience variables and other variables except for sex. This means:

In [73]:
flexible_model_vars = basic_model_vars + ['exp2', 'exp3', 'exp4']

interactions = []
for exp in ['exp1', 'exp2', 'exp3', 'exp4']:
    for var in ['hsg', 'scl', 'clg', 'ad', 'so', 'we', 'ne'] + \
               [col for col in df_with_dummies.columns if col.startswith('occ2_') or col.startswith('ind2_')]:
        interaction_var = df_with_dummies[exp] * df_with_dummies[var]
        interactions.append(interaction_var.values.reshape(-1, 1))

flexible_model_array = np.hstack([df_with_dummies[flexible_model_vars].values] + interactions)

print(flexible_model_array.shape)

(5150, 245)


8.3. The extra-flexible model will include all two-way interactions between variables, except for sex. 

In [74]:
extra_flexible_model_vars = ['exp1', 'exp2', 'exp3', 'exp4', 'hsg', 'scl', 'clg', 'ad', 
                             'so', 'we', 'ne'] + \
                             [col for col in df_with_dummies.columns if col.startswith('occ2_') or col.startswith('ind2_')]

two_way_interactions = []
for i, var1 in enumerate(extra_flexible_model_vars):
    for var2 in extra_flexible_model_vars[i+1:]:
        interaction_var = df_with_dummies[var1] * df_with_dummies[var2]
        two_way_interactions.append(interaction_var.values.reshape(-1, 1))

interactions_array = np.hstack(two_way_interactions)

extra_flexible_model_array = np.hstack([df_with_dummies[extra_flexible_model_vars].values, interactions_array])

print(extra_flexible_model_array.shape)

(5150, 1378)


9.Normalize all the variables, including Y. This means that for each variable V, create V~:

In [75]:
def normalize_array(arr):
    mean = np.mean(arr, axis=0)
    std_dev = np.std(arr, axis=0)
    
    # Prevent SD that are 0
    std_dev[std_dev == 0] = 1  
    
    normalized_array = (arr - mean) / std_dev
    return normalized_array


try:
    basic_model_array = basic_model_array.astype(float)
    flexible_model_array = flexible_model_array.astype(float)
    extra_flexible_model_array = extra_flexible_model_array.astype(float)
    log_w = log_w.astype(float)
    print("All arrays successfully converted to float.")
except ValueError as e:
    print("Error converting to float:", e)



All arrays successfully converted to float.


In [76]:
#For our Y (log_w)
norm_log_w = normalize_array(log_w)

#For basic model
normalized_basic_model_array = normalize_array(basic_model_array)

# For flexible model
normalized_flexible_model_array = normalize_array(flexible_model_array)

# For extra flexible model
normalized_extra_flexible_model_array = normalize_array(extra_flexible_model_array)

## 3. Linear Regressions

10. Split each of the dataframes created (basic, flexible and extra-flexible models) into a training sample (80% of the data) and a test sample. Use the normalized data for this. Hint: You do not need to normalize the data for dummy variables.

In [77]:
# Normalization of only continuos variables 

continuous_vars = ['exp1', 'exp2', 'exp3', 'exp4', 'hsg', 'scl', 'clg', 'ad', 'so', 'we', 'ne']

normalized_continuous_array = normalize_array(df_with_dummies[continuous_vars].values)

dummy_vars = [col for col in df_with_dummies.columns if col.startswith('occ2_') or col.startswith('ind2_')]

#Basic Model
continuous_vars_basic = ['hsg', 'scl', 'clg', 'ad', 'so', 'we', 'ne', 'exp1']
normalized_continuous_basic = normalize_array(df_with_dummies[continuous_vars_basic].values)

normalized_basic_model_array = np.hstack([
    df_with_dummies[['sex']].values,  
    normalized_continuous_basic,       
    df_with_dummies[dummy_vars].values 
])


#Flexible Model 
# Continuous variables for the flexible model
continuous_vars_flexible = continuous_vars_basic + ['exp2', 'exp3', 'exp4']
normalized_continuous_flexible = normalize_array(df_with_dummies[continuous_vars_flexible].values)

interactions = []
for exp in ['exp1', 'exp2', 'exp3', 'exp4']:
    for var in ['hsg', 'scl', 'clg', 'ad', 'so', 'we', 'ne'] + dummy_vars:
        interaction_var = df_with_dummies[exp] * df_with_dummies[var]
        interactions.append(interaction_var.values.reshape(-1, 1))

normalized_flexible_model_array = np.hstack([normalized_continuous_flexible, df_with_dummies[dummy_vars].values] + interactions)

#For the Extra-Flexible Model
normalized_continuous_extra_flexible = normalize_array(df_with_dummies[continuous_vars_flexible].values)

normalized_extra_flexible_model_array = np.hstack([normalized_continuous_extra_flexible, df_with_dummies[dummy_vars].values, interactions_array])


In [78]:
# Import necessary packages 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

b_model = normalized_basic_model_array
f_model = normalized_flexible_model_array
ef_model = normalized_continuous_extra_flexible

y = norm_log_w

In [79]:
b_train,b_test,y_b_train,y_b_test = train_test_split(b_model,y, train_size= 0.8)
f_train,f_test,y_f_train,y_f_test = train_test_split(f_model,y, train_size= 0.8)
ef_train,ef_test,y_ef_train,y_ef_test = train_test_split(ef_model,y, train_size= 0.8)

11. Estimate all three models

- basic
- flexible
- extra-flexible model



In [80]:
#Train

# Basic model 
model_b = LinearRegression()
model_b.fit(b_train,y_b_train)

#Flexible model
model_f = LinearRegression()
model_f.fit(f_train,y_f_train)

#Extra flexible model
model_ef = LinearRegression()
model_ef.fit(ef_train,y_ef_train)



LinearRegression()

In [83]:
#Predict_test
y_b_hat_test = model_b.predict(b_test)
y_f_hat_test = model_f.predict(f_test)
y_ef_hat_test = model_ef.predict(ef_test)

#Predict_train
y_b_hat_train = model_b.predict(b_train)
y_f_hat_train = model_f.predict(f_train)
y_ef_hat_train = model_ef.predict(ef_train)

12. Report the MSE and the R2 for both samples of each model, and the adjusted R2 for the training sample. Do we see overfitting? Which model has the best out-of-sample performance?

In [85]:

# MSE - Out of sample

mse_b_test = np.mean((y_b_test - y_b_hat_test) ** 2)
mse_f_test = np.mean((y_f_test - y_f_hat_test) ** 2)
mse_ef_test = np.mean((y_ef_test - y_ef_hat_test) ** 2)

print(f'MSE out of sample of the basic model is {mse_b_test}')
print(f'MSE out of sample of the flexible model is {mse_f_test}')
print(f'MSE out of sample of the extra flexible is {mse_ef_test}')

# MSE - Sample

mse_b_train = np.mean((y_b_train - y_b_hat_train) ** 2)
mse_f_train = np.mean((y_f_train - y_f_hat_train) ** 2)
mse_ef_train = np.mean((y_ef_train - y_ef_hat_train) ** 2)

print(f'MSE of sample of the basic model is {mse_b_train}')
print(f'MSE of sample of the flexible model is {mse_f_train}')
print(f'MSE of sample of the extra flexible is {mse_ef_train}')

# R-squared - Out of sample

R_sq_1_test = 1 - mse_b_test / np.mean((y_b_test-np.mean(y_b_test))** 2)
R_sq_2_test = 1 - mse_f_test / np.mean((y_f_test-np.mean(y_f_test))** 2)
R_sq_3_test = 1 - mse_ef_test / np.mean((y_ef_test-np.mean(y_ef_test))** 2)

print(f'R-squared out of sample of the basic model is {R_sq_1_test}')
print(f'R-squared out of sample of the flexible model is {R_sq_2_test}')
print(f'R-squared out of sample of the extra flexible model is {R_sq_3_test}')

# R-squared - Sample
R_sq_1_train = 1 - mse_b_train / np.mean((y_b_train-np.mean(y_b_train))** 2)
R_sq_2_train = 1 - mse_f_train / np.mean((y_f_train-np.mean(y_f_train))** 2)
R_sq_3_train = 1 - mse_ef_train / np.mean((y_ef_train-np.mean(y_ef_train))** 2)

print(f'R-squared out of sample of the basic model is {R_sq_1_train}')
print(f'R-squared out of sample of the flexible model is {R_sq_2_train}')
print(f'R-squared out of sample of the extra flexible model is {R_sq_3_train}')

# Adjusted R-squared

n = 5150 
n_sample = n * 0.8

adj_mse_1_train = n_sample / (n_sample-50) * mse_b_train
adjR_sq_1_train = 1 - adj_mse_1_train / np.mean((y_b_train-np.mean(y_b_train))** 2)

adj_mse_2_train = n_sample / (n_sample-245) * mse_f_train
adjR_sq_2_train = 1 - adj_mse_2_train / np.mean((y_f_train-np.mean(y_f_train))** 2)

adj_mse_3_train = n_sample/ (n_sample-1378) * mse_ef_train
adjR_sq_3_train = 1 - adj_mse_3_train / np.mean((y_ef_train-np.mean(y_ef_train))** 2)

print(f'Adjusted R-squared in sample of the basic model is {adjR_sq_1_train}')
print(f'Adjusted R-squared in sample of the flexible model  is {adjR_sq_2_train}')
print(f'Adjusted R-squared in sample of the extra flexible model is {adjR_sq_3_train}')

MSE out of sample of the basic model is 0.6949035167471324
MSE out of sample of the flexible model is 0.7845450292138585
MSE out of sample of the extra flexible is 0.8185159589226129
MSE of sample of the basic model is 0.6902736402438732
MSE of sample of the flexible model is 0.6310864035986009
MSE of sample of the extra flexible is 0.8192693882469149
R-squared out of sample of the basic model is 0.3137175050004656
R-squared out of sample of the flexible model is 0.2918296253373194
R-squared out of sample of the extra flexible model is 0.18148410256207892
R-squared out of sample of the basic model is 0.3074989471507793
R-squared out of sample of the flexible model is 0.3513073491458393
R-squared out of sample of the extra flexible model is 0.18024843561720316
Adjusted R-squared in sample of the basic model is 0.29899156320914266
Adjusted R-squared in sample of the flexible model  is 0.3102932331563504
Adjusted R-squared in sample of the extra flexible model is -0.23172007485671897


These results show that overfitting is present in the extra flexible model. This is visible when we compare the values between its R-squared and adjusted R-squared, as the value reduces drastically from 0.35 to -0.23, which means that the addition of more regressors does not improve the model's fit (a feature of overfitting). In that sense, the model with the best out of sample performance is the basic model, due to having the smallest out of sample MSE value (0.69) and highest out of sample R-squared value (0.31) compared to the other models. 